File: ObjC/BookshelfObjC/DetailViewController.m
//
//  DetailViewController.m
//  BookshelfObjC
//
//  Created by Victor Barros on 2016-03-03.
//  Copyright © 2016 Kinvey. All rights reserved.
//

#import "DetailViewController.h"
@import Kinvey;
@import SVProgressHUD;

@interface DetailViewController ()

@end

@implementation DetailViewController

#pragma mark - Managing the detail item

- (void)setBook:(Book *)book
{
    _book = book;
    
    // Update the view.
    [self configureView];
}

- (void)configureView {
    // Update the user interface for the detail item.
    if (self.book) {
        self.titleTextField.text = self.book.title;
    }
}

-(void)viewDidAppear:(BOOL)animated
{
    [self.titleTextField becomeFirstResponder];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    [self configureView];
}

-(BOOL)shouldPerformSegueWithIdentifier:(NSString *)identifier sender:(id)sender
{
    if ([identifier isEqualToString:@"save"]) {
        if (self.book == nil) {
            self.book = [[Book alloc] init];
        }
        self.book.title = self.titleTextField.text;
        self.book.publicationDate = [NSDate date];
        
        KNVDataStore<Book*>* store = [KNVDataStore getInstance:KNVDataStoreTypeSync
                                                      forClass:[Book class]];
        [SVProgressHUD show];
        [store save:self.book
  completionHandler:^(Book * _Nullable book, NSError * _Nullable error)
        {
            [SVProgressHUD dismiss];
            if (book) {
                [self performSegueWithIdentifier:identifier sender:sender];
            } else {
                UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Error"
                                                                               message:@"Operation not completed"
                                                                        preferredStyle:UIAlertControllerStyleAlert];
                [alert addAction:[UIAlertAction actionWithTitle:@"OK"
                                                          style:UIAlertActionStyleDefault
                                                        handler:nil]];
                [self presentViewController:alert
                                   animated:YES
                                 completion:nil];
            }
        }];
        return NO;
    }
    return YES;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: Swift/Bookshelf/DetailViewController.swift
//
//  DetailViewController.swift
//  Bookshelf
//
//  Created by Victor Barros on 2016-02-08.
//  Copyright © 2016 Kinvey. All rights reserved.
//

import UIKit
import Kinvey
import SVProgressHUD
import MobileCoreServices
import AssetsLibrary
import Photos

class DetailViewController: UIViewController {

    @IBOutlet weak var pictureImageView: UIImageView!
    @IBOutlet weak var titleTextField: UITextField!
    @IBOutlet weak var takeChoosePictureButton: UIButton!

    var store: DataStore<Book>!
    
    var book: Book! {
        didSet {
            // Update the view.
            self.configureView()
        }
    }
    
    func configureView() {
        // Update the user interface for the detail item.
        if let book = book, let titleTextField = titleTextField {
            titleTextField.text = book.title
        }
    }
    
    override func viewDidAppear(animated: Bool) {
        self.titleTextField.becomeFirstResponder()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        self.configureView()
    }
    
    override func shouldPerformSegueWithIdentifier(identifier: String, sender: AnyObject?) -> Bool {
        switch identifier {
        case "save":
            if book == nil {
                book = Book()
            }
            book.title = titleTextField.text
            SVProgressHUD.show()
            
            store.save(book) { (book, error) -> Void in
                SVProgressHUD.dismiss()
                if let _ = book {
                    self.performSegueWithIdentifier(identifier, sender: sender)
                } else {
                    let alert = UIAlertController(title: "Error", message: "Operation not completed", preferredStyle: .Alert)
                    alert.addAction(UIAlertAction(title: "OK", style: .Default, handler: nil))
                    self.presentViewController(alert, animated: true, completion: nil)
                }
            }
            return false
            
        case "cancel":
            if book == nil {
                book = Book()
            }
            
            SVProgressHUD.show()
            if let bookId = book.entityId {
                SVProgressHUD.dismiss()

                //user cancelled, reload book from the cache to disacard any local changes
                store.findById(bookId) { (cachedBook, error) -> Void in
                    if let _ = cachedBook {
                        self.book = cachedBook
                    }
                }
            }
            
            return false
        default:
            return true
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

}

